src/app/clusters/general-commissioning-server/general-commissioning-server.cpp

```diff
@@ -328,23 +521,105 @@ bool emberAfGeneralCommissioningClusterSetRegulatoryConfigCallback(app::CommandH
+    TermsAndConditionsProvider * const termsAndConditionsProvider = Server::GetInstance().GetTermsAndConditionsProvider();
+
+    Optional<TermsAndConditions> acceptedTermsAndConditions = Optional<TermsAndConditions>({
+        .value   = commandData.TCUserResponse,
+        .version = commandData.TCVersion,
+    });
+
+    Commands::SetTCAcknowledgementsResponse::Type response;
+    response.errorCode = CheckTermsAndConditionsAcknowledgementsState(termsAndConditionsProvider, acceptedTermsAndConditions);
+
+    if (CommissioningErrorEnum::kOk == response.errorCode)
+    {
+        CheckSuccess(termsAndConditionsProvider->SetAcceptance(acceptedTermsAndConditions), Failure);
+
+        if (failSafeContext.IsFailSafeArmed())
+        {
+            failSafeContext.SetUpdateTermsAndConditionsHasBeenInvoked();
+        }
+        else
+        {
+            CheckSuccess(termsAndConditionsProvider->CommitAcceptance(), Failure);
+        }
+    }
+
+    commandObj->AddResponse(commandPath, response);
+#endif // CHIP_CONFIG_TC_REQUIRED
+    return true;
+}
+
 namespace {
 void OnPlatformEventHandler(const DeviceLayer::ChipDeviceEvent * event, intptr_t arg)
 {
     if (event->Type == DeviceLayer::DeviceEventType::kFailSafeTimerExpired)
     {
         // Spec says to reset Breadcrumb attribute to 0.
         Breadcrumb::Set(0, 0);
+
+        if (event->FailSafeTimerExpired.updateTermsAndConditionsHasBeenInvoked)
+        {
+#if CHIP_CONFIG_TC_REQUIRED
+            TermsAndConditionsProvider * const termsAndConditionsProvider = Server::GetInstance().GetTermsAndConditionsProvider();
+
+            if (nullptr == termsAndConditionsProvider)
+            {
+                return;
+            }
+
+            // Clear terms and conditions acceptance on failsafe timer expiration
+            termsAndConditionsProvider->RevertAcceptance();
+#endif // CHIP_CONFIG_TC_REQUIRED
+        }
     }
 }
 
 } // anonymous namespace
 
+class GeneralCommissioningFabricTableDelegate : public chip::FabricTable::Delegate
+{
+public:
+    // Gets called when a fabric is deleted
+    void OnFabricRemoved(const FabricTable & fabricTable, FabricIndex fabricIndex) override
+    {
+        // If the FabricIndex matches the last remaining entry in the Fabrics list, then the device SHALL delete all Matter
+        // related data on the node which was created since it was commissioned.
+        if (Server::GetInstance().GetFabricTable().FabricCount() == 0)
+        {
+            ChipLogProgress(Zcl, "general-commissioning-server: Last Fabric index 0x%x was removed",
+                            static_cast<unsigned>(fabricIndex));
+
+#if CHIP_CONFIG_TC_REQUIRED
+            TermsAndConditionsProvider * const termsAndConditionsProvider = Server::GetInstance().GetTermsAndConditionsProvider();
+
+            if (nullptr == termsAndConditionsProvider)
+            {
+                return;
+            }
+
+            termsAndConditionsProvider->ResetAcceptance();
```

Comment:
To be more scalable to future additions, this should just be the `ResetAcceptance` in the `if (nullptr != termsAndConditionsProvider)`